Feature/221011 skill search

beetrbgus opened the pull request:
기술 스택 리스트 추가 및 조회
영일님이 develop에 push 한게 왜 들어가있는지는 모르겠음.

mycode01 committed:
add validator, application exception

mycode01 committed:
remove builder dependency

mycode01 committed:
change wrong error message

mycode01 committed:
add exception handler for MethodArgumentNotValidException

beetrbgus committed:
Skill Dto 및 Entity 관련
feat
- 컬럼 명 명시 및 생성자 추가
- SkillMapper 추가 ( 에러는 없지만 warning 뜨는게 싫어서 id와 imageUrl ignore 처리
style
- SkillDto는 현재 내부 로직에만 사용되기 때문에 Model 패키지로 수정

beetrbgus committed:
Skill Upload 관련

- S3에 업로드시 Key를 받아서 사용할 수 있게 변경
- Skill 등록시 사용할 CreateDto 생성

beetrbgus committed:
Skill 조회 및 등록 관련

- 단일 등록 추가 및 전체 등록 url 수정
- 단일 조회 및 목록 조회 추가

beetrbgus committed:
Skill Entity 및 Dto 관련

- Entity 컬럼 명시 및 @Getter 추가
- Skill 생성시 필요한 RequestDto 추가
- 공통적으로 사용할 Dto 생성
- Mapper 생성 ( Error는 안나는데 warning 뜨는게 싫어서 ignore 추가)

beetrbgus committed:
Skill 조회 및 생성 추가

- ExceptionHandler를 타야하는데 어떤걸 탈지 몰라서 일단 Controller의 28 라인에 IOException 닮 ( Comment 후 수정 예정)
- 단건,전체 조회 및 등록 추가

beetrbgus committed:
Skill 조회 및 생성 추가

- ExceptionHandler를 타야하는데 어떤걸 탈지 몰라서 일단 Controller의 28 라인에 IOException 닮 ( Comment 후 수정 예정)
- 단건,전체 조회 및 등록 추가

beetrbgus committed:
스킬 조회 컨트롤러 수정

beetrbgus committed:
Skill Dto 수정

- 생성자 Builder 추가
- SkillCreateDto -> SkillCreateRequestDto로 변경

beetrbgus committed:
Skill 생성 수정

- SkillCreateDto -> SkillCreateRequestDto로 변경

beetrbgus commented on src/main/java/com/justcodeit/moyeo/study/application/aws/s3/S3Service.java:
```diff
@@ -25,23 +25,22 @@ public class S3Service {
+    public String fileUpload(MultipartFile multipartFile, String folderName) throws IOException{
         String fileName = multipartFile.getOriginalFilename();
-
+        String filePath = folderName+"/"+fileName;
```
S3에 올릴 때 folderName을 적으면 키값에 포함되어서 폴더를 생성

beetrbgus commented on src/main/java/com/justcodeit/moyeo/study/application/skill/SkillService.java:
```diff
@@ -1,21 +1,47 @@
+        return SkillMapper.SKILL_INSTANCE.entityListToSkillList(skillList);
+    }
+    public SkillDto findSkillByName(String name) {
+        Skill skill = skillRepository.findByName(name);
```
Optional로 안한 이유는 값이 없다면
 나중에 ExceptionHandler에서 공통적으로 204 No Content에러로 잡아내기 위해서

mycode01 replied:
Optianal을 사용하든 하지 않든 구현자의 맘대로지만 
204는 에러가 아닙니다.

beetrbgus replied:
그건 맞죵. 원하는거는 결과 없으면 Null을 참조할 때 NullPointException이 뜰텐데 그걸 그냥 204 상태값이랑 메세지로 퉁 치고 싶단 것

mycode01 replied:
이것도 구현하기 나름이긴 한데, 에러가 발생해서 throw가 났을때도 이것을 로직의 분기로 봐야한다/ 아니다 에러는 확실히 따로 분리해서 처리해야한다 가 말이 좀 있습니다.

beetrbgus replied:
'ㅅ' 그럼 만약에 Empty면 어떻게 처리하는게 맞을까요??
Optional 로 뽑았을 때 빈 객체 처리한다고 했을 때
 if(!객체.isPresent) { ?? }

beetrbgus commented on src/main/java/com/justcodeit/moyeo/study/interfaces/resource/SkillController.java:
```diff
@@ -1,20 +1,47 @@
 
     @PostMapping
+    public ResponseEntity<SkillDto> createSkill(@ModelAttribute("skillCreateDto") @Valid SkillCreateRequestDto skillCreateRequestDto) throws IOException {
+        if(!skillCreateRequestDto.isSvg()) {
```
MultiFile 형식은 Valid를 적용할 수 없기 때문에 들어온 값이 svg 파일인지(또는 파일이 없는지)에 대해서 검증로직이 필요

beetrbgus commented on src/main/java/com/justcodeit/moyeo/study/interfaces/resource/SkillController.java:
```diff
@@ -1,20 +1,47 @@
     @PostMapping
+    public ResponseEntity<SkillDto> createSkill(@ModelAttribute("skillCreateDto") @Valid SkillCreateRequestDto skillCreateRequestDto) throws IOException {
+        if(!skillCreateRequestDto.isSvg()) {
+            throw new IOException();
```
에러 코드는 뭘로 해야될지 몰라서 IOException으로 변경 ( 코드 리뷰 때 얘기 필요)

jjoylee commented on src/main/java/com/justcodeit/moyeo/study/interfaces/resource/SkillController.java:
```diff
@@ -1,20 +1,47 @@
+    }
+
+    @PostMapping("/upload")
     public ResponseEntity<Boolean> uploadZip() {
```
저희 void값을 리턴하는 것 아닌가요?

jjoylee commented on src/main/java/com/justcodeit/moyeo/study/interfaces/resource/SkillController.java:
```diff
@@ -1,20 +1,47 @@
+    @PostMapping("/upload")
     public ResponseEntity<Boolean> uploadZip() {
-        return ResponseEntity.ok(skillService.skillSaveAllFromS3());
+        return ResponseEntity.ok(skillService.skillcreateAllFromS3());
```
void 값 return 하면 되지 않나요?

mycode01 replied:
이 api도 왜 있어야하는지 모르겠습니다.
db에 초기값 insert가 필요하다면 dml을 작성해서 초기화 하시기 바랍니다.

beetrbgus replied:
@mycode01 dml이 뭔지 모르기 때문에 그냥 이렇게 짰어요.. 'ㅅ'
@jjoylee 사실 void로 해도 되긴하는데, 상태값 정도는 반환하고 싶어서 줬어요.

mycode01 replied:
그렇게까지 하기 싫다 하시면 뭐 돈받고 하는것도 아닌데 저도 걍 지적질 그만 하겠습니다.

beetrbgus replied:
@mycode01 음 dml 확인요. 일단 의사 전달이 잘못된건지 반응이 부정적인 체념인 것 같아서 말씀드리는데, 
일단 300개 스킬을 전에 말하신 대로 txt 파일로 목록 추출한 다음에 하는 방법도 있긴 하지만 굳이 수작업으로 해야 되는지 의문이 들어서 만든 API입니다. 
dml 뜻 별 생각없이 봐서 이해 못한 부분은 죄송합니다.
하지만 그냥 이건 개인적인 스타일이라고 봐주시면... S3가 담긴 코드 만들어 보고 싶다 정도로 생각해주시면 감사할 것 같습니다.

mycode01 commented on src/main/java/com/justcodeit/moyeo/study/application/aws/s3/S3Service.java:
```diff
@@ -25,23 +25,22 @@ public class S3Service {
 
-    @Value("${moyeo.url.s3EndPoint}")
-    private String S3_ENDPOINT;
+    private final String S3_ENDPOINT = "https://moyeo-skillstack.s3.ap-northeast-2.amazonaws.com/";
```
외부 시스템의 변경이나 환경의 변경에 따라 달라질수 있는 상수는 application.yml같은 환경 파일로 따로 빼주세요.

beetrbgus replied:
@mycode01 저번에 영일님인가..그랬던 것 같아서 일부러 명시한건뎅...일단 확인요.

mycode01 replied:
그때 아마 private final 로 선언하시고 @Value에 상수값 그대로 집어넣으시길래 의도를 명확하게 하라는 취지에서 그런 이야길 했던걸로 기억합니다.

beetrbgus replied:
@mycode01 확인해봤는데 S3Service에서는 S3_ENDPOINT 를 사용할 부분이 없어서 제거하겠습니다.
추가로 ENDPOINT를 사용하는 다른 부분에서는 Value로 설정할게요

jjoylee commented on src/main/java/com/justcodeit/moyeo/study/interfaces/dto/skill/SkillCreateRequestDto.java:
```diff
@@ -0,0 +1,26 @@
+    private String name;
+    private Integer orderNum;
+
+    public boolean isSvg() {
```
이건 개인적으로 dto에 숨기지 않고 서비스 단에서 명시적으로 해줘도 좋을 것 같습니다.

jjoylee commented on src/main/java/com/justcodeit/moyeo/study/application/skill/SkillService.java:
```diff
@@ -1,21 +1,47 @@
+    public SkillDto saveSkill(SkillCreateRequestDto skillCreateRequestDto) throws IOException {
+        String url = s3Service.fileUpload(skillCreateRequestDto.getFile(), skillCreateRequestDto.getFolderName());
+        SkillDto skillDto = SkillMapper.SKILL_INSTANCE.skillDtoToEntity(skillCreateRequestDto);
+        skillDto.setImageUrl(url);
```
개인적으로 SkillCreateRequestDto, entity 만 사용해도 괜찮을 것 같습니다.

yunhozz replied:
저도 같은 생각입니다. 컨트롤러에서 받은 `SkillCreateRequestDto` 를 서비스 레벨에서 변환하지 않고 그대로 사용해도 될 것 같습니다.

beetrbgus replied:
@jjoylee @yunhozz 그럼 궁금한게 있는데 S3에 저장 된 url은 CreateRequestDto가 가지고 있어야 한다고 생각하시나요??
저도 고민을 해봤는데 정작 Request 받을 때에 오지 않는 쓸모없는 변수를 가지고 있는게 아닌가 생각이 들어서 따로 뺀거거든요. Setter는 entity에 사용하지 말자라고 결정했으니까 Entity에 Set 할 방법이 메소드로는 없으니까 imageUrl을 위한 생성자를 만들어야하는건가 싶고 생각 좀 주시면 감사하겠습니다-

jjoylee replied:
@beetrbgus SkillCreateRequestDto를 Skill로 변환한뒤 url set해주면 될 것으로 생각합니다. 저는 setter 자체를 지양하는 것보다 무분별한, 의미를 모르겠는 setter를 지양하자는 의미로 저는 생각합니다. 아니면 정말 생성자를 만들어서 new Skill(url, CreateRequestDto) 이런 식으로 설정해도 좋을 것 같습니다.

mycode01 commented on src/main/java/com/justcodeit/moyeo/study/application/skill/SkillService.java:
```diff
@@ -1,21 +1,47 @@
 import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
 
+@Transactional
```
Transactional 어노테이션의 사용법에 맞게 사용해주세요

beetrbgus replied:
?? 추상적이라 이해가 안되는데 자세하게 말씀 좀 부탁드릴게요

yunhozz replied:
클래스 위에 `@Transactional` 을 사용하는 것도 딱히 문제가 없다고 생각합니다. 다만, 읽기 전용 메소드에 `@Transactional(readOnly=true)` 를 추가해야 성능상 이점을 얻을 수 있을 것 같습니다.

beetrbgus replied:
아 전 반대로 읽기 전용 메소드에 리드 온리 붙이려고 했는데 무튼 확인용

mycode01 replied:
@yunhozz 클래스 위에 Transactional을 사용하게 되면 그 안에 모든 메소드가 트랜잭션이 걸린 상태로 동작하게 됩니다. 제공하는 대부분의 메소드가 public인데, 굳이 트랜잭션과 관련없는 부분에 의도치 않게 트랜잭션이 걸리게 되고(다른 트랜잭션의 일부로 사용되는 경우), 
또 일부 트랜잭션 구간안에서도 트랜잭션을 따로 걸어야하는 상황에서 그게 여의치 않게 될거같네요.

yunhozz replied:
@mycode01 오호.. 다른 트랜잭션의 일부로 사용되는 경우를 미처 생각을 못했네요.

beetrbgus replied:
@mycode01 굳이 쓸데없는 부분까지 트랜잭션을 걸면 성능상에 문제가 있을 수도 있으니 메소드 단위로 걸라는 말씀인거죠??

mycode01 commented on src/main/java/com/justcodeit/moyeo/study/application/skill/SkillService.java:
```diff
@@ -1,21 +1,47 @@
+        skillRepository.save(skill);
+
+        return SkillMapper.SKILL_INSTANCE.entityToSkillDto(skill);
+    }
```
컨벤션에 맞게 코드를 작성해주세요. 문서 포메팅좀 부탁합니다.

beetrbgus replied:
?? 추상적이라 이해가 안되는데 어떤 부분인지 자세하게 말씀 좀 부탁드릴게요22

mycode01 replied:
커밋하기전에 문서 포메팅 버튼 한번만 눌러주세요.

beetrbgus replied:
@mycode01 음 인텔리 제이에 코드 정렬 단축키를 말씀하시는건가요??

mycode01 commented on src/main/java/com/justcodeit/moyeo/study/interfaces/resource/SkillController.java:
```diff
@@ -1,20 +1,47 @@
     private final SkillService skillService;
 
     @PostMapping
+    public ResponseEntity<SkillDto> createSkill(@ModelAttribute("skillCreateDto") @Valid SkillCreateRequestDto skillCreateRequestDto) throws IOException {
```
이게 필요한 api인지 모르겠습니다.
서비스에 필요한 api도 아니고, 어뷰징에 안전한 기능도 아닙니다.

beetrbgus replied:
관리자 페이지 만들면 필요한 기능이라 당장은 필요없긴 한데 그냥 만들었습니당.
어뷰징 방지를 위해서 @Secure 설정으로  ROLE 기반으로 작성하긴 할건데 제 쪽에서 아직 유저 로그인 쪽 없어서 토큰 값 못받아서 @Secure 작성 안했습니다. Todo 달고 완료될 때 할게요

mycode01 commented on src/main/java/com/justcodeit/moyeo/study/persistence/Skill.java:
```diff
@@ -20,11 +24,27 @@ public class Skill {
     private String imageUrl;
 
+    @Column(name = "order_num")
+    private Integer orderNum;
```
궁금한게 있는데 int가 아니라 굳이 Integer를 사용한 이유는 ordering priority 에 중립적인 skill이 있어도 된다고 판단하셔서 그런건가요?

beetrbgus replied:
현재 우리 프로젝트에서는 몇개의 Skill만 우선순위를 갖고 있고 그 외는 우선순위가 없습니다.
우선순위가 없는 것들을 999이런식으로 해도 되긴하지만 그건 좀 아닌 것 같아서 DB에 저장 될 필요가 없기 때문에 orderNum을 Integer로 작성했습니다.

yunhozz commented on src/main/java/com/justcodeit/moyeo/study/interfaces/mapper/SkillMapper.java:
```diff
@@ -0,0 +1,23 @@
+    SkillMapper SKILL_INSTANCE = Mappers.getMapper(SkillMapper.class);
+
+    SkillDto entityToSkillDto(Skill skill);
+    Skill skillDtoToEntity(SkillDto skillDto);
+    @Mapping(target = "id", ignore = true)
+    @Mapping(target = "imageUrl", ignore = true)
+    SkillDto skillDtoToEntity(SkillCreateRequestDto skillCreateRequestDto);
```
반환 타입은 다르지만 메소드 명이 같아서 혼란이 있을 것 같습니다.

beetrbgus replied:
메소드명 변경 확인

yunhozz commented on src/main/java/com/justcodeit/moyeo/study/persistence/Skill.java:
```diff
@@ -20,11 +24,27 @@ public class Skill {
+    @Column(name = "order_num")
+    private Integer orderNum;
+
+    @Builder
+    public Skill(SkillCategory skillCategory, String folderName, String name, String imageUrl, Integer orderNum) {
+        this.skillCategory = skillCategory;
+        this.folderName = folderName;
+        this.name = name;
+        this.imageUrl = imageUrl;
+        this.orderNum = orderNum;
+    }
```
`@Builder` 만 사용하실거면 private 로 제한하는 것이 나을 것 같습니다.

beetrbgus replied:
프라이빗하면 MapStruct 생성할 때 참조가 안될 것 같은데..확인해보고 말씀드릴게요.

yunhozz replied:
어차피 빌더 자체가 static class라 상관 없지 않을까용??

beetrbgus replied:
@yunhozz 으흠- 변경해도 빌드할 때 정상적으로 작동하네요. 오호...감사합니다. 혹시 참조할 만한 글이 있을까요

beetrbgus committed:
사용하지 않는 Endpoint 제거

- S3 EndPoint 제거

beetrbgus committed:
Entity 및 Dto 변경

- Builder public -> private으로 변경
- imageUrl을 CreateDto에 추가 ( CreateDto -> SkillDto -> Skill로 변환하는 중간 과정 줄임)

beetrbgus committed:
SkillService Refactoring

Service 변경
CreateDto -> SkillDto -> Skill 로 변환하는 중간 과정 생략

Mapstruct 변경
- CreateDto -> SkillDto -> Skill 로 변환하는 중간 과정 제거하면서 사용하지 않는 메서드 삭제
